dj/sql/parsing/ast.py

```diff
@@ -220,25 +230,61 @@ class Expression(Node):
     """an expression type simply for type checking"""
 
 
+@dataclass(eq=False)
+class Name(Node):
+    """
+    the string name specified in sql with quote style
+    """
+
+    name: str
+    quote_style: str = ""
+
+    def __str__(self) -> str:
+        return (
+            f"{self.quote_style}{self.name}{self.quote_style}"  # pylint: disable=C0301
+        )
+
+    def to_identifier(self) -> "Identifier":
+        """
+        transforms a single Name to a single item Identifier
+        """
+        return cast(Identifier, Identifier([self]).add_self_as_parent())
+
+    def __hash__(self) -> int:
+        return hash(self.name + self.quote_style)
+
+
+@dataclass(eq=False)
+class Identifier(Node):
+    """
+    Represents a sequence of names for something Named in a query
+    """
+
+    idents: List[Name]
```

Comment:
Doesn't have to be in this PR but would it make sense for us to add helper methods right on this Identifier class to help find different elements within the identifier?

For example `i.database()`, `i.namespace()`, `i.table()`, `i.node()` (last two might be the same logic). For something like `foo.bar.baz`, `i.catalog()` would return `foo` and for `bar.baz`, `i.catalog()` would return `None`. That kind of stuff.

Reply:
Part of the reason I would argue against this is because it is sort of escaping some other niceties of the AST like there being nodes you can rely on for each entity you would find in a query. You might see what I'm getting at with `Column` having a `_table` attribute and `table` property but that's as far as I went and didn't assume anything else about a database besides columns belong to tables. The idea there is that some downstream validation happens which resolves whether the name is a table or something, adds a table node after examining the column identifier and changes the column identifier to be just the column name. 

I was hesitant to suggest that everything would always be something like database.catalog/schema.table or table.column because I was imagining you could have something like `basic.metrics.users.num_users` as a dj node if I'm not mistaken but there's not really a database, catalog, etc there. Finally, I'm not sure what the ansi spec says about this or how databases would handle this, but sqloxide seems to have no problem dragging things out such as `"select a.b.c.d.e.f.g from x"` so it seems from that there could be arbitrarily long "compound identifiers" that don't necessarily match any schema like the database.catalog....

Reply:
That makes sense, I had the same hesitation. It's a huge pain that it's not really standardized anywhere and here we have the additional pattern of `[namespace_part_a, namespace_part_b, ..., dj_node_name]` that we'd have to account for. You've convinced me it's better to leave that out of the AST.